### ISSUE TITLE
[Question] Hot-reload managed assemblies using assembly file lock instead of through FileStream

### ISSUE THREAD

**Issue by PickleUpADoodle** [natemcmaster/DotNetCorePlugins#291]
I have a question about the methods available for hot-reloading managed DLL files. 

This library only supports hot-reload by loading assembly files in memory through a FileStream. Is the only reason locks cannot be used the fact that a loaded assembly cannot be removed from disk, thus making it not possible for the "changed" event to happen regarding the FileWatcher? Or are there more complications making it not possible/feasible? Locks in this case referring to the loading methods where "load-by-name" or path is used (not file stream).

**Comment by aloksharma1** [comment]
@PickleUpADoodle yes you are right track, if you want hot reloads you need to use shadow caching (keeping loaded copy separate from disk) this provides lock-freed dlls that can be replace on runtime and trigged with file watcher others you will get access denied errors as file will be always in use.

**Comment by DMiradakis** [comment]
@aloksharma1 yep, shadow copying is exactly what I am targeting for my dotnet job orchestrator [Didact](https://github.com/DidactHQ/didact). The entire background job system is built from loading class libraries as dynamic, runtime, post-app-startup plugins, and this is the only option that I saw. Glad to know others have come to the same conclusions.
